File: js/queue.js
module.exports = (function () {
    function Queue(list) {
        if (list === void 0) { list = new Array(); }
        this._list = list.slice();
    }
    Object.defineProperty(Queue.prototype, "list", {
        get: function () {
            return this._list;
        },
        enumerable: true,
        configurable: true
    });
    Queue.prototype.peek = function () {
        return this._list[0];
    };
    Queue.prototype.enqueue = function (value) {
        this._list.push(value);
    };
    Queue.prototype.dequeue = function () {
        return this._list.splice(0, 1);
    };
    Queue.prototype.delete = function (value, count) {
        if (count === void 0) { count = 1; }
        for (var i = 0; i < count; i++) {
            var index = this._list.indexOf(value);
            if (index >= 0 && index < this._list.length)
                this._list.splice(index, 1);
        }
    };
    return Queue;
}());

File: C#/Dex/Dex/Queue.cs
using System.Collections.Generic;

namespace Dex {
	public class Queue<T> {
		public List<T> List { get; set; }

		public Queue(List<T> list = null) {
			this.List = list == null ? new List<T>() : new List<T>(list);
		}

		// Returns the next element in the queue without removing it
		public T Peek() {
			return this.List[0];
		}

		// Adds an element to the end of the queue
		public void Enqueue(T value) {
			this.List.Add(value);
		}

		// Returns the next element in the queue
		public T Dequeue() {
			if (this.List.Count > 0) {
				var value = this.List[0];
				this.List.RemoveAt(0);
				return value;
			} else {
				return default(T);
			}
		}

		// Deletes the specified value from the queue the specified number of times
		public void Delete(T value, int count = 1) {
			for (int i = 0; i < count; i++) {
				int index = this.List.IndexOf(value);

				if (index >= 0 && index < this.List.Count)
					this.List.RemoveAt(index);
			}
		}
	}
}